### ISSUE TITLE
[$500] Thread - Unable to send a reply in thread after leaving the thread in offline mode

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#34424]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:**
**Slack conversation:**

## Action Performed:
1. Go to workspace chat
2. Go offline.
3. Send a message and reply in thread.
4. Leave the thread.
5. Open the thread and send a reply again.

## Expected Result:
User is able to send a reply in thread after leaving the thread in offline mode.

## Actual Result:
User is unable to send a reply in thread after leaving the thread in offline mode.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/43996225/ef933503-fb1c-49ff-9e79-296f3b4ee5ea

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0190e8891c6eb5ad7c</li>
        <li>Upwork Job ID: 1745748146336444416</li>
        <li>Last Price Increase: 2024-01-26</li>
<li>Automatic offers: </li>
<ul>
<li>mollfpr | Reviewer | 28143176</li>
<li>tienifr | Contributor | 28143177</li>
</ul></li>
    </ul>
</details>

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Thread - Unable to send a reply inside of a thread after leaving the thread in offline mode

### What is the root cause of that problem?

The [`canLeaveRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/libs/ReportUtils.ts#L3843) function is used in the [`HeaderView`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L117) component to display the leave room menu option.


https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166-L171

I believe solely using [`isWorkspaceMemberLeavingWorkspaceRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166) to create a condition in the [`leaveRoom`](https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2074) function for this PR https://github.com/Expensify/App/pull/33053 was not a good idea. 

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2092-L2097

This is because PR removes the `report.ReportID` from the onyx if [`isWorkspaceMemberLeavingWorkspaceRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166) is `false`. 
Hence, threads are also affected.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L250-L251

Therefore, `isReportReadyForDisplay`  will always be `false`, and the `ReportScreen` will load infinitely.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L517-L518

It is just not displaying the `ReportActionsView` component.

Finally, the user can not reply because the `DragAndDropProvider` component is disabled.

### What changes do you think we should make in order to solve the problem?




#### Part 1

We will use the `isThread` function with the `isWorkspaceMemberLeavingWorkspaceRoom` variable to create a better condition.

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2092

Modify the above condition to something similar to the one below:

```
value: isWorkspaceMemberLeavingWorkspaceRoom || isThread(report)
```

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2152-L2154

#### Part 2
Finally, we create a new condition after `API.write` shown above. 
It will navigate the user back to the parent report. 

```
if (isThread(report)) {
    Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute(report.parentReportID));
}
```

**Question**
Is navigating back part not supposed to happen for threads? 🤔 
I just added it in case.

**Result** (macOS - Chrome)
https://github.com/Expensify/App/assets/22982173/ba321858-6fd0-49de-8e87-2ecb0eda39ae

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Thread - Unable to send a reply inside of a thread after leaving the thread in offline mode

### What is the root cause of that problem?

The [`canLeaveRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/libs/ReportUtils.ts#L3843) function is used in the [`HeaderView`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L117) component to display the leave room menu option.


https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166-L171

I believe solely using [`isWorkspaceMemberLeavingWorkspaceRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166) to create a condition in the [`leaveRoom`](https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2074) function for this PR https://github.com/Expensify/App/pull/33053 was not a good idea. 

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2092-L2097

This is because PR removes the `report.ReportID` from the onyx if [`isWorkspaceMemberLeavingWorkspaceRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166) is `false`. 
Hence, threads are also affected.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L250-L251

Therefore, `isReportReadyForDisplay`  will always be `false`, and the `ReportScreen` will load infinitely.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L517-L518

It is just not displaying the `ReportActionsView` component.

Finally, the user can not reply because the `DragAndDropProvider` component is disabled.

### What changes do you think we should make in order to solve the problem?




#### Part 1

We will use the `isThread` function with the `isWorkspaceMemberLeavingWorkspaceRoom` variable to create a better condition.

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2092

Modify the above condition to something similar to the one below:

```
value: isWorkspaceMemberLeavingWorkspaceRoom || isThread(report)
```

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2152-L2154

#### Part 2
Finally, we create a new condition after `API.write` shown above. 
It will navigate the user back to the parent report. 

```
if (isThread(report)) {
    Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute(report.parentReportID));
}
```

**Question**
Is the navigating back part not supposed to happen for threads? 🤔 
I added part 2 in case it was needed.

**Result (macOS - Chrome)**

https://github.com/Expensify/App/assets/22982173/fe8288c8-90cd-4014-845b-6843a49dc150

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Thread - Unable to send a reply inside of a thread after leaving the thread in offline mode

### What is the root cause of that problem?

The [`canLeaveRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/libs/ReportUtils.ts#L3843) function is used in the [`HeaderView`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L117) component to display the leave room menu option.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166-L171

I believe solely using [`isWorkspaceMemberLeavingWorkspaceRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166) to create a condition in the [`leaveRoom`](https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2074) function for this PR https://github.com/Expensify/App/pull/33053 was not a good idea. 

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2092-L2097

This is because PR removes the `report.ReportID` from the onyx if [`isWorkspaceMemberLeavingWorkspaceRoom`](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/HeaderView.js#L166) is `false`. 
Hence, threads are also affected.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L250-L251

Therefore, `isReportReadyForDisplay`  will always be `false`, and the `ReportScreen` will load infinitely.

https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L517-L518

It is just not displaying the `ReportActionsView` component.

Finally, the user can not reply because of the `ReportFooter` component over [here](https://github.com/Expensify/App/blob/a2864a077fd5d577021983ff77c658c9f2e43bab/src/pages/home/ReportScreen.js#L546).

### What changes do you think we should make in order to solve the problem?


#### Part 1

We will use the `isThread` function with the `isWorkspaceMemberLeavingWorkspaceRoom` variable to create a better condition.

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2092

Modify the above condition to something similar to the one below:

```
value: isWorkspaceMemberLeavingWorkspaceRoom || isThread(report)
```

https://github.com/Expensify/App/blob/f268c39c331305c49dbb80a0d0e59984289ecb25/src/libs/actions/Report.ts#L2152-L2154

#### Part 2
Finally, we create a new condition after `API.write` shown above. 
It will navigate the user back to the parent report. 

```
if (isThread(report)) {
    Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute(report.parentReportID));
}
```

**Question**
Is the navigating back part not supposed to happen for threads? 🤔 
I added part 2 in case it was needed.

**Result (macOS - Chrome)**

https://github.com/Expensify/App/assets/22982173/fe8288c8-90cd-4014-845b-6843a49dc150

**Comment by joekaufmanexpensify** [comment]
Pending proposal reviews from @thesahindia

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by joekaufmanexpensify** [comment]
TY!

**Comment by mollfpr** [comment]
Sorry @joekaufmanexpensify, I misunderstood the issue here. 

I just checked when I was online. After leaving the thread and sending a message to the thread, I will re-join the thread and send the message.

The expected behavior I mentioned before is showing the skeleton loader after leaving the thread while offline. For reference https://github.com/Expensify/App/issues/28997 and https://expensify.slack.com/archives/C01GTK53T8Q/p1699960963387459

**Comment by mollfpr** [comment]
>We should not delete the thread report/ report action after leaving thread, we just need to update the notification preference to not receive the notification.

@tienifr Could you point out which line deleting the report action offline?

**Comment by mollfpr** [comment]
@Tony-MK I don't see you test it offline in your video.

Also, what's the root cause for the user can't send the message?

**Comment by Tony-MK** [comment]
My apologies, @mollfpr, for uploading the wrong recording. 
Here is an accurate recording.

https://github.com/Expensify/App/assets/22982173/2b64eda5-8d48-4074-a9fe-8623756e709f


> Also, what's the root cause for the user can't send the message?

The root cause is the same. 
It is because the `report.reportID` is not present. 
Therefore, `isReportReadyForDisplay` will always be false.
Hence, the `handleSendMessage` quits before sending the function because of the condition seen below.

https://github.com/Expensify/App/blob/599c005849cd065a53d35e015991b637c768f2da/src/pages/home/report/ReportActionCompose/ReportActionCompose.js#L343-L355

**Comment by joekaufmanexpensify** [comment]
Still pending further input

**Comment by mollfpr** [comment]
> Is the navigating back part not supposed to happen for threads?

@Tony-MK I'm not sure how to answer this, but the current behavior is navigating back the user to the parent report.

>We will use the isChatThread function with the isWorkspaceMemberLeavingWorkspaceRoom variable to create a better condition.

https://github.com/Expensify/App/blob/5edf172457de71f5fa77d9ddf30e03726db346a2/src/pages/home/HeaderView.js#L164

https://github.com/Expensify/App/blob/5edf172457de71f5fa77d9ddf30e03726db346a2/src/libs/actions/Report.ts#L2107-L2116

The condition `isWorkspaceMemberLeavingWorkspaceRoom` is checked from `!isChatThread`, so basically, the second condition will always skipped.

---

>We should not delete the thread report/ report action after leaving thread, we just need to update the notification preference to not receive the notification.

@tienifr Does this make the issue https://github.com/Expensify/App/issues/33010 resurface again? Also, we have the `stateNum` and `statusNum` values updated on optimistic. I'm not sure if this will lead to a regression if we remove it.

**Comment by Tony-MK** [comment]
@mollfpr, can you reproduce the bug?

It seems like it is fixed.

**Comment by Tony-MK** [comment]
@mollfpr, can you reproduce the bug?

It is partially fixed.

> @Tony-MK I'm not sure how to answer this, but the current behavior is navigating back the user to the parent report.

However, it does not navigate back to the parent report.

**Comment by joekaufmanexpensify** [comment]
Solutions still being discussed

**Comment by mollfpr** [comment]
@Tony-MK Yes.

>It is partially fixed.

Which part is fixed?

>However, it does not navigate back to the parent report.

It's still navigating back to the parent report.


https://github.com/Expensify/App/assets/25520267/6cf0197b-1623-4995-9d41-0e024bac7ca6


---

@tienifr I think we don't need to update `stateNum` and `statusNum optimistically`. After the thread leaves, the API `OpenReport` doesn't update that data.

https://github.com/Expensify/App/blob/9c32b8fe5642ac766b910bd3078061980087b202/src/libs/actions/Report.ts#L2175-L2183

How about the above code? Does the logic be implemented for nonworkspace rooms?

**Comment by joekaufmanexpensify** [comment]
@mollfpr are we close to selecting a proposal here?

**Comment by mollfpr** [comment]
I think the [proposal](https://github.com/Expensify/App/issues/34424#issuecomment-1888814273) from @tienifr will work. Checking on the response from API when opening a thread after leaves, it only updates the `notificationPreference`. Let's go with their proposal.

🎀 👀 🎀 C+ reviewed!

**Comment by joekaufmanexpensify** [comment]
Pending internal sign off

**Comment by joekaufmanexpensify** [comment]
@grgia mind signing off on this one when you have a sec? TY!

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by joekaufmanexpensify** [comment]
DM'ed Georgia about proposal

**Comment by grgia** [comment]
All yours @tienifr

**Comment by joekaufmanexpensify** [comment]
@grgia mind confirming [this question](https://github.com/Expensify/App/pull/35869#issuecomment-1938786610)? Seems like the PR review is blocked on it.

**Comment by mollfpr** [comment]
We have another PR that fixed this issue, and the solution is kinda similar. I have checked the PR https://github.com/Expensify/App/pull/37355, and it fixes this issue.

I think we can close the PR, but I don't know what the next process.

cc @joekaufmanexpensify

**Comment by joekaufmanexpensify** [comment]
Got it. I'll take a look.

**Comment by joekaufmanexpensify** [comment]
[Discussing ](https://expensify.slack.com/archives/C01SKUP7QR0/p1709679648479319)

**Comment by joekaufmanexpensify** [comment]
We discussed and landed on the following:

- @tienifr - $500 for contributor role, since you were hired for this job, wrote the PR, and the delay was on us internally. 
- @mollfpr - $250 for C+, since you reviewed proposals, but did not do a full review.

**Comment by joekaufmanexpensify** [comment]
@tienifr $500 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
@mollfpr please accept our offer so we can pay you.

**Comment by joekaufmanexpensify** [comment]
I'm OOO the next two days, so we'll issue payment to @mollfpr on Monday (once you accept the offer).

**Comment by mollfpr** [comment]
Sorry for the late reply @joekaufmanexpensify I did manual request in NewDot.

**Comment by joekaufmanexpensify** [comment]
Ah, all good. I canceled the offer in upwork, and updated my post above to clarify that you're being paid via NewDot.

**Comment by joekaufmanexpensify** [comment]
All set!

**Comment by JmillsExpensify** [comment]
$250 approved for @mollfpr based on [this summary](https://github.com/Expensify/App/issues/34424#issuecomment-1979886131).
